### PR TITLE
SN1-447: add logits

### PR DESCRIPTION
**Pull request by Hollyqui**

### CODE REVIEW

File: prompting/rewards/exact_match.py [hunk 1]
```diff
@@ -27,72 +34,159 @@ def normalize_timing(timing: float, timings: float) -> float:
...
+                for idx in verify_indices:
+                    check_idx = min(idx, completion_length - 1)
+                    if not chunk_dicts_raw[check_idx].choices[0].logprobs:
+                        # logger.debug(f"NO LOGPROBS FOR CHUNK: {chunk_dicts_raw[check_idx]}")
+                        # logger.debug(f"LOGPROBS: {chunk_dicts_raw[check_idx].choices[0].logprobs}")
+                        verification_scores.append(0.0)
+                        continue
+
+                    if chunk_dicts_raw[check_idx].choices[0].logprobs.content is None:
+                        logger.warning(f"Miner failed to provide logits: {chunk_dicts_raw[check_idx]}")
```
Comment (dbobrenko) [inline prompting/rewards/exact_match.py:147]: We should lower log level to `debug`

File: prompting/rewards/exact_match.py [hunk 2]
```diff
@@ -27,72 +34,159 @@ def normalize_timing(timing: float, timings: float) -> float:
...
+                        verification_scores.append(0.0)
+                        continue
+                    original_logits = {
+                        info.token: info.logprob
+                        for info in chunk_dicts_raw[check_idx].choices[0].logprobs.content[0].top_logprobs
+                    }
+
+                    verification_output, prompt = await self.model_manager.get_model(task.llm_model_id).generate_logits(
```
Comment (dbobrenko) [inline prompting/rewards/exact_match.py:155]: How fast is it to perform one reward iteration for 100 miners?

Also will it reload the model if it rotates to another `model_id` in the middle of check?

File: prompting/rewards/exact_match.py [hunk 3]
```diff
@@ -27,72 +34,159 @@ def normalize_timing(timing: float, timings: float) -> float:
...
+        sampling_parameters: dict = task.sampling_params
+        PENALIZE_ALL = BatchRewardOutput(
+            rewards=np.array([-INCORRECT_PENALTY] * len(completions)),
+            timings=np.array([0.0] * len(completions)),
+        )
+
+        max_length = 0
+        for chunk in all_chunks:
+            if chunk and max_length < len(chunk):
+                max_length = len(chunk)
```
Comment (dbobrenko) [inline prompting/rewards/exact_match.py:106]: Shouldn't we take max_length and sample indices for each miner individually?

Assume one miner submits response with 1M tokens -> we sample 100k indices in range(0, 1M). Then we will do only single check for the last logit for all other miners with high probability

File: prompting/rewards/exact_match.py [hunk 4]
```diff
@@ -27,72 +34,159 @@ def normalize_timing(timing: float, timings: float) -> float:
...
+                    continue
+
+                # Verify logits for selected indices
+                verification_scores = []
+                completion_length = len(chunks)
+
+                for idx in verify_indices:
+                    check_idx = min(idx, completion_length - 1)
+                    if not chunk_dicts_raw[check_idx].choices[0].logprobs:
+                        # logger.debug(f"NO LOGPROBS FOR CHUNK: {chunk_dicts_raw[check_idx]}")
```
Comment (dbobrenko) [inline prompting/rewards/exact_match.py:141]: Let's remove all non-used commented out code

**Review by dbobrenko — COMMENTED** [review]
Added some more comments

File: prompting/rewards/exact_match.py [hunk 5]
```diff
@@ -27,72 +34,159 @@ def normalize_timing(timing: float, timings: float) -> float:
...
+            timings=np.array([0.0] * len(completions)),
+        )
+
+        max_length = 0
+        for chunk in all_chunks:
+            if chunk and max_length < len(chunk):
+                max_length = len(chunk)
+
+        if max_length == 0:
+            logger.debug("NO CHUNKS TO VERIFY, PENALIZING ALL")
```
Comment (dbobrenko) [inline prompting/rewards/exact_match.py:109]: Why screaming :joy:

File: prompting/llms/hf_text_image.py [hunk 6]
```diff
@@ -6,25 +6,25 @@
...
-            device_map=self._device,
-        )
-        self.model = self.model.to(self._device)
-        self.tokenizer = AutoProcessor.from_pretrained(model_id)
-        self.valid_generation_params = set(self.model.generation_config.to_dict().keys())
+from prompting.llms.vllm_llm import ReproducibleVLLM
+
+
+class VLLMTextImageToText(ReproducibleVLLM):
+    # def __init__(
```
Comment (dbobrenko) [inline prompting/llms/hf_text_image.py:13]: Let's remove init if it's no longer needed

File: prompting/llms/model_manager.py [hunk 7]
```diff
@@ -222,6 +222,8 @@ async def start(self, scoring_queue: list, name: str | None = None, **kwargs):
 
     async def run_step(self):
         """This method is called periodically according to the interval."""
+        if self.llm_model_manager.active_models:
+            self.interval = 600
```
Comment (dbobrenko) [inline prompting/llms/model_manager.py:226]: Why do we need this?

File: prompting/llms/model_manager.py [hunk 8]
```diff
@@ -213,7 +213,7 @@ async def _vram_cleanup(self):
 
 class AsyncModelScheduler(AsyncLoopRunner):
     llm_model_manager: ModelManager
-    interval: int = 14400
+    interval: int = 10
```
Comment (dbobrenko) [inline prompting/llms/model_manager.py:216]: Make sure to restore original value before merge

File: prompting/tasks/task_creation.py [hunk 9]
```diff
@@ -20,7 +20,7 @@
 class TaskLoop(AsyncLoopRunner):
     is_running: bool = False
     thread: threading.Thread = None
-    interval: int = 0
+    interval: int = 20
```
Comment (dbobrenko) [inline prompting/tasks/task_creation.py:23]: Do we want to increase latency for task sending?

File: shared/logging/logging.py [hunk 10]
```diff
@@ -186,16 +186,46 @@ class RewardLoggingEvent(BaseEvent):
...
+            "reward_events": self.reward_events,
+            "task_id": self.task_id,
+            "task": self.task,
+            "task_dict": self.task_dict,
+            "source": self.source,
+            "reference": self.reference,
+            "challenge": self.challenge,
+        }
+
+    # def __str__(self):
```
Comment (dbobrenko) [inline shared/logging/logging.py:218]: Ditto

File: prompting/llms/model_manager.py (hunk 8 above)
Comment (bkb2135) [reply prompting/llms/model_manager.py:216]: We override this later to be every 10 minutes, 14400 seconds seems like a long time to rotate

File: prompting/llms/model_manager.py (hunk 7 above)
Comment (bkb2135) [reply prompting/llms/model_manager.py:226]: So our first load in happens quickly, but once we load at least one model in the loops will slow down

File: prompting/llms/hf_text_image.py (hunk 6 above)
Comment (bkb2135) [reply prompting/llms/hf_text_image.py:13]: This class is still WIP, the whole class is unused, we will need to check with multimodal and stuff

File: prompting/rewards/exact_match.py (hunk 5 above)
Comment (bkb2135) [reply prompting/rewards/exact_match.py:109]: I'm mad there are no chunks

File: scripts/install.sh [hunk 11]
```diff
@@ -7,8 +7,8 @@ pip install poetry
 poetry config virtualenvs.in-project true
 
 # Install the project dependencies
-poetry install --extras "validator"
-poetry run pip install flash-attn --no-build-isolation
+poetry install --extras "validator" --no-cache
+poetry run pip install vllm
```
Comment (dbobrenko) [inline scripts/install.sh:11]: I think it's worth to bind it to a specific version to avoid sudden surprises. Also can we move it to poetry?

**Review by dbobrenko — APPROVED** [review]
LGTM
